Sources/MapboxCoreNavigation/EHorizon/RoadObjectDistanceInfo.swift

```diff
@@ -0,0 +1,134 @@
+     */
+    case polygon(identifier: RoadObjectIdentifier,
+                 type: RoadObjectType,
+                 distanceToNearestEntry: CLLocationDistance,
+                 distanceToNearestExit: CLLocationDistance,
+                 isInside: Bool)
+
+    /**
+     The information about distance to the road object represented as a subgraph.
+     - parameter identifier: Road object identifier.
+     - parameter type: Road object type.
+     - parameter distanceToNearestEntry: Distance measured in meters to the nearest entry.
+     - parameter distanceToNearestExit: Distance measured in meters to the nearest exit.
+     - parameter isInside: Boolean that indicates whether we're currently "inside" the object.
+     */
+    case subgraph(identifier: RoadObjectIdentifier,
+                  type: RoadObjectType,
+                  distanceToNearestEntry: CLLocationDistance,
+                  distanceToNearestExit: CLLocationDistance,
+                  isInside: Bool)
+    
+    /**
+     The information about distance to the road object represented as a line.
+     - parameter identifier: Road object identifier.
+     - parameter type: Road object type.
+     - parameter distanceToEntry: Distance from the current position to entry point measured in meters along the road graph. This value is 0 if already "within" the object.
+     - parameter distanceToExit" Distance from the current position to the most likely exit point measured in meters along the road graph.
+     - parameter distanceToEnd: Distance from the current position to the most distance exit point measured in meters along the road graph.
+     - parameter isEntryFromStart: Boolean that indicates whether we enter the road object from its start. This value is `false` if already "within" the object.
+     - parameter length: Length of the road object measured in meters.
+     */
+    case line(identifier: RoadObjectIdentifier,
+              type: RoadObjectType,
+              distanceToEntry: CLLocationDistance,
+              distanceToExit: CLLocationDistance,
+              distanceToEnd: CLLocationDistance,
+              isEntryFromStart: Bool,
+              length: CLLocationDistance)
+
+    /** Road object identifier */
+    public var roadObjectIdentifier: RoadObjectIdentifier {
+        switch self {
+        case .point(let identifier, _, _),
+             .gantry(let identifier, _, _),
+             .polygon(let identifier, _, _, _, _),
+             .subgraph(let identifier, _, _, _, _),
+             .line(let identifier, _, _, _, _, _, _):
+            return identifier
+        }
+    }
+
+    /** Road object type */
+    public var roadObjectType: RoadObjectType {
+        switch self {
+        case .point(_, let type, _),
+             .gantry(_, let type, _),
+             .polygon(_, let type, _, _, _),
+             .subgraph(_, let type, _, _, _),
+             .line(_, let type, _, _, _, _, _):
+            return type
+        }
+    }
```

Comment:
These two fields are common among each of the cases, so they don’t need to be part of the enumeration but could instead live one level up in the object model, where they’d be more accessible. It looks like only a notification user info dictionary uses RoadObjectDistanceInfo, storing them in an array. Ideally, the user info dictionary would go back to putting RoadObjectDistanceInfo in a dictionary, but the keys would be instances of a type similar to RoadObject (but without a `length` and `location`) or the former RoadObjectDistance (but without the `distanceInfo` property).

Reply:
In my opinion, these properties demonstrate that it’s a bit awkward to include the identifier and type as associated values, but I don’t see a great alternative. It’s an unfortunate consequence of the underlying model that MapboxNavigationNative exposes.

Reply:
As long as we leave RoadObjectDistanceInfo as a flat enumeration, we should find a more descriptive name for it, such as DistancedRoadObject.